modules/scala-http4s/src/main/scala/dev/guardrail/generators/scala/http4s/Http4sServerGenerator.scala

```diff
@@ -206,18 +206,16 @@ class Http4sServerGenerator private (version: Http4sVersion)(implicit Cl: Collec
         schemes.foldM[F, Either[$authErrorTypeName, $authContextTypeName]](Left($errorTermName.Unauthorized)){ case (result, el) => 
           result match {
             case Left(value) => 
-              el.toNel.map({
-                case (scheme, scopes) =>
-                  middleware(scheme, scopes, req)
-              }).reduce[F[Either[$authErrorTypeName, $authContextTypeName]]]({
-                case (l, r) =>
-                  l.flatMap({
-                    case Right(_) =>
-                      r
-                    case l: Left[$authErrorTypeName, $authContextTypeName] =>
-                      Applicative[F].pure(l)
-                  })
-              })
+              val nel = el.toNel
```

Comment:
One caveat of the `NonEmptyMap` is that it seems to preclude the ["Optional Oauth2 Security"](https://swagger.io/specification/#security-requirement-object) example, reproduced here:

```yaml
security:
  - {}
  - petstore_auth:
    - write:pets
    - read:pets
```

This looks like an oversight from the first foray into authentication, so if you'd like to consider that case out of scope here I'm happy to take that on, if you're running out of time. I appreciate the number of iterations you've taken to this problem so far, just trying to make sure we merge something that isn't going to need more work immediately.

Reply:
that's true, I've wrote about the bug above. I would fix it before merging the change because it's not very useful with the problem

Reply:
ah, I've checked the link and it's not the bug I mentioned, it's another use case related with optional requirements, anyway, I think it needs to be fixed)

Reply:
as I wrote in the previous comment, the swagger library we use under the hood doesn't support this approach to define optional authentication. It seems there is the `x-security-optional` to cover the case, but it was a list of names which is a bit overwhelming and unclear. To make it a bit more clear, I've changed its type to Boolean and it could be defined just for a route to make its security requirements optional.
For the `custom` authentication we provide its value to the `authenticationMiddleware` and let the user to decide what to do with that: produce some special case of its `AuthContext` or just pass it through inside of the `AuthContext` to the handler and deal with that there.
For the `simple` authentication we replace `Either` by an `Option` for the handler but the `authenticationMiddleware`  signature wasn't changed